### PR TITLE
Align URLSearchParams with the spec

### PR DESCRIPTION
**Pull request by JustOff**
See https://url.spec.whatwg.org/#interface-urlsearchparams

Based on:
 - https://bugzilla.mozilla.org/show_bug.cgi?id=1330678
 - https://bugzilla.mozilla.org/show_bug.cgi?id=1330698
 - https://bugzilla.mozilla.org/show_bug.cgi?id=1330699
 - https://bugzilla.mozilla.org/show_bug.cgi?id=1331580
 - https://bugzilla.mozilla.org/show_bug.cgi?id=1366032

*Warning, the tests have not been ported*.

This requires additional thorough checks, due to the large number of complex changes.

Tag #1001.

### CODE REVIEW

File: dom/url/URLSearchParams.cpp [hunk 1]
```diff
@@ -335,34 +327,43 @@ URLSearchParams::WrapObject(JSContext* aCx, JS::Handle<JSObject*> aGivenProto)
...
+    }
+  } else if (aInit.IsUSVStringUSVStringRecord()) {
+    const Record<nsString, nsString>& record =
+      aInit.GetAsUSVStringUSVStringRecord();
+    for (auto& entry : record.Entries()) {
+      sp->Append(entry.mKey, entry.mValue);
+    }
   } else {
-    sp->ParseInput(input);
+    MOZ_CRASH("This should not happen.");
```
Comment (wolfbeast) [inline dom/url/URLSearchParams.cpp]: Nit: let's avoid Vague and Undetermined crash error messages.
Suggestion: "URLSearchParams: Invalid string"

Comment (JustOff) [reply dom/url/URLSearchParams.cpp]: Fixed.

**Comment by g4jc** [comment]
@JustOff FYI, I am getting a build error in master as of yesterday, appears to be related to these rename commits that just landed.
```
21:08.98 In file included from /uxp/obj-x86_64-pc-linux-gnu/dist/include/mozilla/dom/Headers.h:10:0,
21:08.98                  from /uxp/obj-x86_64-pc-linux-gnu/dist/include/mozilla/dom/WorkerScope.h:12,
21:08.98                  from /uxp/dom/bindings/BindingUtils.cpp:53,
21:08.98                  from /uxp/obj-x86_64-pc-linux-gnu/dom/bindings/Unified_cpp_dom_bindings0.cpp:2:
21:08.99 /uxp/obj-x86_64-pc-linux-gnu/dist/include/mozilla/dom/HeadersBinding.h:13:10: fatal error: mozilla/dom/Record.h: No such file or directory
21:08.99  #include "mozilla/dom/Record.h"
21:08.99           ^~~~~~~~~~~~~~~~~~~~~~
21:08.99 compilation terminated.

```

**Comment by JustOff** [comment]
@g4jc Everything builds fine for me on Windows, perhaps you need to try `mach clobber`?

**Comment by g4jc** [comment]
Works again for me after a git --hard reset and clobber, must have been some file got stuck in limbo. Thanks!

**Pull request merged**